ethpmV3 support

njgheorghita opened the pull request:
Truffle support for the ethpmV3 spec!

Hey @gnidan and @eggplantzzz . Thanks for your help so far in discussing this. Here's a first pass at ethpmV3 x truffle. I'm sure there's plenty of areas where my javascript is smelly, please point them out anywhere you see them. After a review, I plan to run this through another thorough round of integration tests, & I have an update to the `trufflesuite.com` docs in the pipeline to reflect these changes.

njgheorghita commented on packages/core/lib/commands/packages.js:
```diff
@@ -0,0 +1,24 @@
+var command = {
```
We discussed separating the `truffle packages` command out into it's own pr, so that we could discuss its api in greater depth, however for the sake of getting an mvp in and across the line, I've left it in here. Happy to remove it out into it's own pr, but it's unclear right now how much more time I have available to extend the `packages` api.

njgheorghita commented on packages/config/src/configDefaults.ts:
```diff
@@ -33,11 +33,13 @@ export const getInitialConfig = ({
+      ipfsProtocol: "https",
+      ipfsPort: "5001",
+      registry: {
+        address: "0x8011df4830b4f696cd81393997e5371b93338878",
```
I'm happy to deploy an un-authenticated registry for truffle to use as its default, unless you guys want to handle that. I'm 50/50 on whether it should be on mainnet (better reliability / authority / helps minimize spam packages) or on any testnet (much much cheaper / easier to start getting your hands dirty with publishing ethpm packages)

gnidan replied:
does this field respect ENS names?

njgheorghita replied:
Yep, though right now it only permits ens on mainnet. However, upon further investigation it looks like truffle supports ens on all testnets except kovan, so I'll make the update to suppport ens on these other testnets.

njgheorghita commented on packages/ethpm-v1/README.md:
````diff
@@ -0,0 +1,9 @@
+const ethpmV1 = require('ethpm-v1');
+```
+
+This package contains support for the deprecated v1 version of the ethpm specification for backwards compatibility purposes.
````
Just as a heads-up, I haven't done any editing / testing to the v1 code. It was just copied into its own package from the initial implementation inside the core package. The v1 tests were (and still are) set to be skipped.

njgheorghita commented on packages/ethpm-v3/lib/ethpm-v3.js:
```diff
@@ -0,0 +1,485 @@
+      gasPrice: 100000000000,
+      data: encodedTxData
+    });
+    try {
```
It seems like there might be a better mechanism available somewhere in truffle that manages sending txs with a more user-friendly interface? Could you point me in the right direction...

njgheorghita commented on packages/core/lib/commands/install.js:
```diff
@@ -3,33 +3,43 @@ const command = {
     ]
   },
-  run: function(options, done) {
+  run: async function (options, done) {
```
When testing locally, I occasionally run into edge cases that hang indefinitely. Is there an existing mechanism to handle these cases? Would it be a good idea to implement some kind of timeout limit when running an ethpm command?

gnidan replied:
This might be because you're making a callback style function also async. Generally not a good idea to do that - use one or the other (see other comment about `callbackify`)

gnidan commented on packages/core/lib/commands/install.js:
```diff
@@ -3,33 +3,43 @@ const command = {
   builder: {},
   help: {
-    usage: "truffle install <package_name>[@<version>]",
+    usage: "truffle install <packageId> [--alias]",
```
```suggestion
    usage: "truffle install <package-identifier> [--alias]",
```

gnidan commented on packages/core/lib/commands/install.js:
```diff
@@ -3,33 +3,43 @@ const command = {
-        option: "package_name",
-        description:
-          "Name of the package as listed in the Ethereum Package Registry. (required)"
+        option: "packageId",
```
```suggestion
        option: "packageIdentifier",
```

gnidan commented on packages/core/lib/commands/install.js:
```diff
@@ -3,33 +3,43 @@ const command = {
 
-    if (options._ && options._.length > 0) options.packages = options._;
+    if (options._ && options._.length == 0) {
+      done(new Error(`Please provide a packageId.`));
```
```suggestion
      done(new Error(`Please provide a package identifier.`));
```

Should also probably expand on this error with additional info, like, "Please see `truffle help install` for more information on valid package identifiers."

gnidan commented on packages/core/lib/commands/install.js:
```diff
@@ -3,33 +3,43 @@ const command = {
+    if (options._ && options._.length > 1) {
+      done(new Error(`Only one packageId can be installed at a time.`));
+    }
+    options.packageId = options._[0];
```
```suggestion
    options.packageIdentifier = options._[0];
```

gnidan commented on packages/core/lib/commands/install.js:
```diff
@@ -3,33 +3,43 @@ const command = {
-      })
-      .catch(done);
+
+    if (config.ethpm.version == "1") {
```
The ideal requirement I have in mind here is that `truffle install` will install whatever package, whether it's v1 or v3. Was this not possible? Are we forced to make the user decide between one or the other for installation? (I know that it's required to make this determination ahead of time for publish)

njgheorghita replied:
Yeah, automatic version detection of manifests is not so straightforward, there are a lot of edge cases possible here. One thing to keep in mind is that the read api differs between the v1 registry and v3 registries. I feel we should encourage users to only use v3 (except for installing existing v1 packages). And in that case, they should be encouraged to update their packages to v3, and republish.

I guess this brings up another thought, which is adding a deprecation notice whenever users use v1. Do you have any objections to that?

eggplantzzz replied:
I don't have any problem with a deprecation notice. @gnidan?

gnidan commented on packages/core/lib/commands/packages.js:
```diff
@@ -0,0 +1,24 @@
+
+    var config = Config.detect(options);
+
+    if (config.ethpm.version == "1" || config.ethpm.version == "3") {
```
Does PackageV3 support listing packages for v1?

njgheorghita replied:
Yeah good catch, it doesn't, but I've gone ahead and updated `ethpm-v1` to mimic the behavior in `ethpm-v3` to display info about the release on the v1 ethpm registry.

gnidan commented on packages/core/lib/commands/publish.js:
```diff
@@ -6,16 +6,21 @@ var command = {
     options: []
   },
-  run: function(options, done) {
+  run: async function (options, done) {
```
Oh I should note that it's not wise to make a function like this be `async` and take a callback. I'd suggest instead using `callbackify` (`const { callbackify } = require("util")`) and not bothering with the callback arg at all. See [Node `callbackify` docs](https://nodejs.org/api/util.html#util_util_callbackify_original)

gnidan commented on packages/core/lib/commands/packages.js:
```diff
@@ -0,0 +1,24 @@
+    usage: "truffle packages",
+    options: []
+  },
+  run: async function (options, done) {
```
(async functions shouldn't take callbacks also... one or the other, else things can break in subtle ways)

eggplantzzz commented on packages/core/lib/commands/install.js:
```diff
@@ -1,35 +1,54 @@
-        description:
-          "Name of the package as listed in the Ethereum Package Registry. (required)"
+        option: "packageIdentifier",
+        description: `(required) Name of the package as listed in the Ethereum Package Registry. Accepted formats: packageName, packageName@version, ethpm URI, ipfs URI.`
```
It is convention in Truffle to put the "(required)" at the end of the description.

eggplantzzz commented on packages/core/lib/commands/install.js:
```diff
@@ -1,35 +1,54 @@
   builder: {},
   help: {
-    usage: "truffle install <package_name>[@<version>]",
+    usage: "truffle install <package-identifier> [--alias]",
```
Looks like convention in Truffle is snake case when specifying options of this format. So it should be `package_identifiers`.

njgheorghita replied:
:+1: I've made this change - but just want to note that `package-identifier` was suggested above by @gnidan.

eggplantzzz commented on packages/core/lib/commands/install.js:
```diff
@@ -1,35 +1,54 @@
-        option: "package_name",
-        description:
-          "Name of the package as listed in the Ethereum Package Registry. (required)"
+        option: "packageIdentifier",
```
Same as the above comment

njgheorghita replied:
:+1: I've made this change - but just want to note that `packageIdentifier` was suggested above by @gnidan.

eggplantzzz replied:
Can you still make this change? It should be snake case.

eggplantzzz commented on packages/core/lib/commands/install.js:
```diff
@@ -1,35 +1,54 @@
+    let callbackFunction;
+
+    if (config.ethpm.version == "1") {
+      callbackFunction = callbackify(PackageV1.install);
```
This should probably be named something like `install` or `installPackage` as it is really not a callback function.

eggplantzzz commented on packages/core/lib/commands/packages.js:
```diff
@@ -0,0 +1,30 @@
+    options: []
+  },
+  run: function (options, done) {
+    var Config = require("@truffle/config");
```
Another small thing: could you rename these `var`s `const`? Anywhere there is a variable that we won't be setting again it should be `const`. `var`s are kinda old school

eggplantzzz commented on packages/core/lib/commands/publish.js:
```diff
@@ -6,16 +8,22 @@ var command = {
   },
-  run: function(options, done) {
+  run: function (options, done) {
     var Config = require("@truffle/config");
```
Same with these guys.

eggplantzzz commented on packages/core/lib/commands/publish.js:
```diff
@@ -6,16 +8,22 @@ var command = {
-        return done();
-      })
-      .catch(done);
+    let callbackFunction;
```
Same as the above about the name of this function.

eggplantzzz commented on packages/ethpm-v1/lib/ethpm-v1.js:
```diff
@@ -1,16 +1,76 @@
+        `Please include a provider in your networks config for the Ropsten testnet, the testnet where the ethpmV1 registry is deployed.`
+      );
+    }
+    const provider = options.networks["ropsten"].provider();
```
Will this always be a method? I feel like this property is usually accessed like `options.networks[<name>].provider`. I could be wrong, I'll have to check this out.

eggplantzzz commented on packages/ethpm-v1/lib/ethpm-v1.js:
```diff
@@ -1,16 +1,76 @@
+    const provider = options.networks["ropsten"].provider();
+    const web3 = new Web3(provider);
+    const ethpmV1RegistryAddress = "0x8011df4830b4f696cd81393997e5371b93338878";
+    var contract = new web3.eth.Contract(
```
Do we want to use `@truffle/contract` instead of `web3` for this?

eggplantzzz commented on packages/ethpm-v3/lib/utils.js:
```diff
@@ -0,0 +1,301 @@
+  const sourcePaths = {};
+
+  // group artifacts by network id to ensure consistency when converting networkId => blockchainUri
+  for (let file of files) {
```
You can use `const` (like `for (const file of files) {`) here as well.

eggplantzzz commented on packages/ethpm-v3/lib/utils.js:
```diff
@@ -0,0 +1,301 @@
+      if (options.networks[provider].network_id == networkId) {
+        // handle ganache provider from tests
+        if (
+          options.networks[provider].provider.constructor.name == "Function"
```
It is probably better to use `===` which is strict equals.

You could also do 
```
if (typeof options.networks[provider].provider === "function") {
```
if you like but I suppose your way is good too. Actually, how is this indexed by `provider`? Is that really some provider value?

njgheorghita replied:
I've updated it to your way - this was a bit of a hack to handle ganache provider instances in the tests, while also supporting users setting provider methods as provider in `truffle-config`. Though (as in the other provider-related comment) I could be misunderstanding how users typically set their provider in `truffle-config`.

eggplantzzz commented on packages/ethpm-v3/lib/utils.js:
```diff
@@ -0,0 +1,301 @@
+
+async function resolveEnsName(address, provider, options) {
+  let resolvedAddress;
+  const w3 = new Web3(provider);
```
Convention in Truffle is to use `web3` as the variable name for the instance of `Web3`.

eggplantzzz commented on packages/ethpm-v3/lib/utils.js:
```diff
@@ -0,0 +1,301 @@
+    return {};
+  }
+  const installedBuildDependencies = {};
+  Object.keys(ethpmLock).forEach(function (key, _) {
```
You can just omit the 2nd argument (`_`) in this case.

An arrow function would be slickest.
```
Object.keys(ethpmLock).forEach(key => installedBuildDependencies[key] = ethpmLock[key].resolved_uri)
```

njgheorghita commented:
@eggplantzzz Thanks for the review! I've changed or responded to your comments. Let me know if there's anything else glaring / obvious that needs addressing. In terms of the failing tests... Has there been any significant changes to the resolver recently? It seems to me that the `await Migrate.run(config)` step is failing in the tests - and it's not obvious the best way to fix it, but I'll be able to dig a little deeper into it tomorrow if it's not obvious to you what changed.

eggplantzzz commented:
Sorry about jerking you back and forth on the `package-identifier` stuff, I didn't see @gnidan's suggestion about it until afterwards haha.

eggplantzzz commented:
As for the CI failures, I don't think anything changed with the migrations. I'll take a quick look and see if I can find anything out about this.

eggplantzzz commented:
So @njgheorghita, I took a look and your tests are failing because there were a bunch of breaking changes to some of the compilation packages. `workflow-compile` had compilation and the saving of artifacts separated into two steps. I made the necessary corrections and pushed them to a branch in the Truffle repo called `fix-ethpm-tests`. I didn't want to push directly to your branch but feel free to merge it into yours or fix it yourself. The gist of the matter is that `Contracts.compile` no longer saves the artifacts. There is now a method `Contracts.compileAndSave` that you need to use to do both. We also started using `WorkflowCompile` as the name when importing that library and I made that change in there as well. Let me know if something is unclear or if you need anything else!

njgheorghita commented:
@eggplantzzz Nailed it! Thanks for looking into it and the fix!

njgheorghita commented:
@eggplantzzz I can definitely update `packageIdentifier` -> `package_identifier`. Just to make sure I'm not messing up anything here - what's the reason to merge the `develop` branch into this one? Is it fine to continue building on top of it?

njgheorghita commented:
I've also updated the ethpm x truffle docs [here](https://github.com/trufflesuite/trufflesuite.com/pull/620) and added my blog (which is fairly similar to the docs but has a better example and more in-depth analysis) [here](https://github.com/trufflesuite/trufflesuite.com/pull/817).

eggplantzzz commented:
@njgheorghita Oh, there was just a merge conflict so I resolved it for you. I hope you don't mind. And thanks for the fix!
